syntax = "proto3";

package fileservice;

service Fileservice {
	
	rpc UploadFile (FileData) returns (ack);
	rpc DownloadFile (FileInfo) returns (FileData);
	rpc FileSearch (FileInfo) returns (ack);
	rpc ReplicateFile (stream FileData) returns (ack);
	rpc FileList (UserInfo) returns (FileListResponse);
	rpc FileDelete (FileInfo) returns (ack);
	rpc UpdateFile (stream FileData) returns (ack);
	rpc getClusterStats (Empty) returns (ClusterStats);
  	rpc getLeaderInfo (ClusterInfo) returns (ack);
	rpc MetaDataInfo (MetaData) returns (ack);
}

message FileData {
	string username = 1;
	string filename =2;
	bytes data =3;
	string message = 4;
}

message MetaData {
	string filename = 1;
	bytes seqValues = 2;
}

message ack {
	bool success =1;
	string message = 2;
}

message UserInfo {
	string username=1;
}

message FileInfo {
	string username = 1;
	string filename = 2;
}

message FileListResponse {
	string Filenames =1;
}

message ClusterInfo {
  string ip = 1;
  string port = 2;
  string clusterName = 3;
}

message Empty {}

message ClusterStats {
   
  string cpu_usage = 1;
  string disk_space = 2; 
  string used_mem = 3;
}
